### PR TITLE
Warn users when they have invalid format

### LINKED ISSUE #1985: Produce a warning when spaces have invalid formats
## Motivation 

There is problem https://github.com/tarantool/tarantool/issues/7841

There is code fix: https://github.com/tarantool/tarantool/wiki/Fix-illegal-field-type-in-a-space-format-when-upgrading-to-2.10.4

## Task

The task is to adopt code fix to produce warning in cartridge.

### PR DESCRIPTION
**Pull request by yngvar-antonsson**
I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #1985

### CODE REVIEW

File: cartridge/invalid-format.lua [hunk 1]
```diff
@@ -0,0 +1,62 @@
+local log = require('log')
+
+local illegal_types = {
+    [''] = true,
+    ['n'] = true,
+    ['nu'] = true,
+    ['s'] = true,
+    ['st'] = true,
+}
```
Comment (filonenko-mikhail) [inline cartridge/invalid-format.lua:9]: very good

**Review by filonenko-mikhail — APPROVED** [review]

File: cartridge/invalid-format.lua [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+local function before_replace(_, tuple)
+    if tuple == nil then return tuple end
+
+    local name = tuple[3]
+    local format = tuple[7]
+
+    for _, field_def in ipairs(format) do
+        if illegal_types[field_def.type] then
+            _G.__cartridge_invalid_format_spaces[name] = true
+        end
```
Comment (Totktonada) [inline cartridge/invalid-format.lua:23]: I would also look at `field_def[2]`. AFAIU a user may have a space created, when the old format representation was the default one.

File: cartridge/invalid-format.lua [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+-- _space trigger.
+local function before_replace(_, tuple)
+    if tuple == nil then return tuple end
+
+    local name = tuple[3]
+    local format = tuple[7]
+
+    for _, field_def in ipairs(format) do
+        if illegal_types[field_def.type] then
+            _G.__cartridge_invalid_format_spaces[name] = true
```
Comment (Totktonada) [inline cartridge/invalid-format.lua:22]: Is it verified with `strict.on()` / on a Debug build? I see `rawset()` above, so existence of the global variable is not tracked and so assigning to it in the script mode will raise an error.

The easier way to solve it is to assign the global directly:

```lua
_G.__cartridge_invalid_format_spaces = {}
```

It is allowed on the module level.

The alternative is to make all accesseses via `rawget()`.

(However, please, don't lean on my words and verify it.)

BTW, why not use a module-local variable?

File: cartridge/invalid-format.lua [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            return
+        end
+        box.space._space:before_replace(nil, before_replace)
+        box.ctl.on_schema_init(nil, on_schema_init)
+
+        if next(_G.__cartridge_invalid_format_spaces) then
+            log.warn(
+                "You have spaces with invalid format: " ..
+                spaces_name_to_str() ..
+                ". Fix it before use Tarantool 2.10.4 - it's deprecated and " ..
```
Comment (Totktonada) [inline cartridge/invalid-format.lua:60]: BTW, the trigger allows to bootstrap on tarantool 2.10.4 as well.

File: cartridge/confapplier.lua [hunk 5]
```diff
@@ -609,6 +612,8 @@ local function boot_instance(clusterwide_config)
         end
     end
 
+    invalid_format.end_check()
```
Comment (Totktonada) [inline cartridge/confapplier.lua:615]: We definitely can exit from the function before this statement. Is it crucial?

File: cartridge/invalid-format.lua [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    end,
+    end_check = function()
+        if box.ctl.on_schema_init == nil then
+            return
+        end
+        box.space._space:before_replace(nil, before_replace)
+        box.ctl.on_schema_init(nil, on_schema_init)
+
+        if next(_G.__cartridge_invalid_format_spaces) then
+            log.warn(
```
Comment (Totktonada) [inline cartridge/invalid-format.lua:57]: Shouldn't we put it into issues that are shown in the web interface?

File: cartridge/invalid-format.lua (hunk 3 above)
Comment (yngvar-antonsson) [reply cartridge/invalid-format.lua:22]: @Totktonada 
To the last question - I used it in tests only. So, I can move it under the test flag, that we've used before in our code. 
And all the tests run with `strict.on()`.

File: cartridge/invalid-format.lua (hunk 6 above)
Comment (yngvar-antonsson) [reply cartridge/invalid-format.lua:57]: Cool idea!

Comment (Totktonada) [inline cartridge/invalid-format.lua:57]: The issue (#1985) asks to migrate the bad formats. Is there implecit agreement to issue a warning instead?

BTW, the issue will not be shown on tarantool 2.10.4. We'll get a fail on bootstrap. Just noted. Don't know, whether it is intentional.

File: cartridge/invalid-format.lua (hunk 3 above)
Comment (Totktonada) [reply cartridge/invalid-format.lua:22]: Hm. Okay. Maybe I don't know something about the strict mode. If it works, disregard my comment.

**Review by Totktonada — APPROVED** [review]
I don't have enough experience to actually review cartridge's code. I left some comments and, I guess, I have nothing more to add. I'll approve, but, please, don't take my approve here with any level of trust.

**Pull request merged**